crates/oxc_linter/src/rules/react/no_array_index_key.rs

```diff
@@ -0,0 +1,305 @@
+        };
+
+        if ident.name.as_str() != prop_name {
+            return;
+        }
+
+        let Some(JSXAttributeValue::ExpressionContainer(container)) = &attr.value else {
+            return;
+        };
+
+        let JSXExpression::Identifier(expr) = &container.expression else {
+            return;
+        };
+
+        if expr.name.as_str() == index_param_name {
+            ctx.diagnostic(no_array_index_key_diagnostic(attr.span));
+        }
+    }
+}
+
+fn check_react_clone_element<'a>(
+    call_expr: &'a CallExpression,
+    node: &'a AstNode,
+    ctx: &'a LintContext,
+) {
+    let Some(index_param_name) = find_index_param_name(node, ctx) else {
+        return;
+    };
+
+    if is_method_call(call_expr, Some(&["React"]), Some(&["cloneElement"]), Some(2), Some(3)) {
+        let Some(Argument::ObjectExpression(obj_expr)) = call_expr.arguments.get(1) else {
+            return;
+        };
+
+        for prop_kind in &obj_expr.properties {
+            let ObjectPropertyKind::ObjectProperty(prop) = prop_kind else {
+                continue;
+            };
+
+            let PropertyKey::StaticIdentifier(key_ident) = &prop.key else {
+                continue;
+            };
+
+            let Expression::Identifier(value_ident) = &prop.value else {
+                continue;
+            };
+
+            if key_ident.name.as_str() == "key" && value_ident.name.as_str() == index_param_name {
+                ctx.diagnostic(no_array_index_key_diagnostic(obj_expr.span));
+            }
+        }
+    }
+}
+
+fn find_index_param_name<'a>(node: &'a AstNode, ctx: &'a LintContext) -> Option<&'a str> {
+    for ancestor in ctx.nodes().iter_parents(node.id()).skip(1) {
+        if let AstKind::CallExpression(call_expr) = ancestor.kind() {
+            let Expression::StaticMemberExpression(expr) = &call_expr.callee else {
+                return None;
+            };
+
+            if SECOND_INDEX_METHODS.contains(expr.property.name.as_str()) {
+                return find_index_param_name_by_position(call_expr, 1);
+            }
+
+            if THIRD_INDEX_METHODS.contains(expr.property.name.as_str()) {
+                return find_index_param_name_by_position(call_expr, 2);
+            }
+        }
+    }
+
+    None
+}
+
+fn find_index_param_name_by_position<'a>(
+    call_expr: &'a CallExpression,
+    position: usize,
+) -> Option<&'a str> {
+    match &call_expr.arguments[0] {
+        Argument::ArrowFunctionExpression(arrow_fn_expr) => {
+            return Some(
+                arrow_fn_expr.params.items.get(position)?.pattern.get_identifier()?.as_str(),
+            );
+        }
+
+        Argument::FunctionExpression(regular_fn_expr) => {
+            return Some(
+                regular_fn_expr.params.items.get(position)?.pattern.get_identifier()?.as_str(),
+            );
+        }
+
+        _ => (),
+    }
+
+    None
+}
+
+const SECOND_INDEX_METHODS: phf::Set<&'static str> = phf::phf_set! {
```

Comment:
we could probably just make this a single map of method name to index, but this is all good for now - thanks!